File: Tries/cpp/Auto Complete.cpp
/*
Auto complete

Given n number of words and an incomplete word w. You need to auto-complete that word w.
That means, find and print all the possible words which can be formed using the incomplete word w.
Note : Order of words does not matter.

Input Format :
The first line of input contains an integer, that denotes the value of n.
The following line contains n space separated words.
The following line contains the word w, that has to be auto-completed.

Output Format :
Print all possible words in separate lines.

Constraints:
Time Limit: 1 sec

Sample Input 1 :
7
do dont no not note notes den
no

Sample Output 2 :
no
not
note
notes

Sample Input 2 :
7
do dont no not note notes den
de

Sample Output 2 :
den

Sample Input 3 :
7
do dont no not note notes den
nom

Sample Output 3 :
(Empty) There is no word which starts with "nom"
*/

#include <bits/stdc++.h>
using namespace std;
class TrieNode
{
public:
  char data;
  TrieNode **children;
  bool isTerminal;

  TrieNode(char data)
  {
    this->data = data;
    children = new TrieNode *[26];
    for (int i = 0; i < 26; i++)
    {
      children[i] = NULL;
    }
    isTerminal = false;
  }
};

class Trie
{
  TrieNode *root;

public:
  int count;

  Trie()
  {
    this->count = 0;
    root = new TrieNode('\0');
  }

  void insertWord(TrieNode *root, string word)
  {
    if (word.length() == 0)
    {
      root->isTerminal = true;
    }
    else
    {
      int index = word[0] - 'a';
      TrieNode *child = root->children[index];
      if (child == NULL)
      {
        child = new TrieNode(word[0]);
        root->children[index] = child;
      }
      insertWord(child, word.substr(1));
    }
  }

  void insertWord(string word)
  {
    insertWord(this->root, word);
  }

  void print(TrieNode *root, string op)
  {
    if (root->isTerminal == true)
    {
      cout << op << "\n";
    }

    for (int i = 0; i < 26; i++)
    {
      if (root->children[i] != NULL)
      {
        TrieNode *child = root->children[i];
        print(child, op + child->data);
      }
    }
  }

  void helper(TrieNode *root, string pattern, string op)
  {
    if (pattern.length() == 0)
    {
      // pattern found
      print(root, op);
    }
    else
    {
      int index = pattern[0] - 'a';
      TrieNode *child = root->children[index];
      if (child != NULL)
      {
        op.push_back(pattern[0]);
        helper(child, pattern.substr(1), op);
      }
    }
  }

  void autoComplete(vector<string> input, string pattern)
  {
    // find if pattern even exists
    for (string s : input)
    {
      insertWord(s);
    }

    helper(root, pattern, "");
  }
};

File: Tries/java/Auto Complete.java
/*
Auto complete

Given n number of words and an incomplete word w. You need to auto-complete that word w.
That means, find and print all the possible words which can be formed using the incomplete word w.
Note : Order of words does not matter.

Input Format :
The first line of input contains an integer, that denotes the value of n.
The following line contains n space separated words.
The following line contains the word w, that has to be auto-completed.

Output Format :
Print all possible words in separate lines.

Constraints:
Time Limit: 1 sec

Sample Input 1 :
7
do dont no not note notes den
no

Sample Output 2 :
no
not
note
notes

Sample Input 2 :
7
do dont no not note notes den
de

Sample Output 2 :
den

Sample Input 3 :
7
do dont no not note notes den
nom

Sample Output 3 :
(Empty) There is no word which starts with "nom"
*/

import java.util.*;
import java.util.ArrayList;

class TrieNode{
    char data;
    boolean isTerminating;
    TrieNode children[];
    int childCount;

    public TrieNode(char data) {
        this.data = data;
        isTerminating = false;
        children = new TrieNode[26];
        childCount = 0;
    }
}

public class Trie {
    private TrieNode root;
    public int count;
    public Trie() {
        root = new TrieNode('\0');
    }

    private void add(TrieNode root, String word){
        if(word.length() == 0){
            root.isTerminating = true;
            return;
        }		
        int childIndex = word.charAt(0) - 'a';
        TrieNode child = root.children[childIndex];
        if(child == null){
            child = new TrieNode(word.charAt(0));
            root.children[childIndex] = child;
            root.childCount++;
        }
        add(child, word.substring(1));
    }

    public void add(String word){
        add(root, word);
    }

    void print(TrieNode root, StringBuilder op){
        if(root.isTerminating == true){
            System.out.println(op);
        }
        for(int i=0; i<root.children.length; i++){
            if(root.children[i] != null){
                TrieNode child = root.children[i];
				StringBuilder temp = new StringBuilder(op);
                temp.append(child.data);
                print(child, temp);
            }
        }
    }

    void helper(TrieNode root, String pattern, StringBuilder op){
        if(pattern.length() == 0){
            // Pattern found OR Pattern is empty
            print(root, op);
            return;
        }
        int index = pattern.charAt(0) - 'a';
        TrieNode child = root.children[index];
        if(child != null){
            op.append(pattern.charAt(0));
            helper(child, pattern.substring(1), op);
        }
    }

    public void autoComplete(ArrayList<String> input, String pattern) {
        for(String s: input){
            add(s);
        }

        StringBuilder s = new StringBuilder("");
        helper(root, pattern, s);
    }

}